### PR TITLE
Saturating Math for i32/integers

### PR DESCRIPTION
**Pull request by thatzopoulos**
Issue : https://github.com/timescale/timescaledb-toolkit/issues/423

This PR adds saturating math functions for addition/subtraction/multiplication for i32.

Each math operator has two functions: 

1. `*_saturate_to_min` saturates at the numeric bounds instead of overflowing, similar to Rusts builtin saturating methods.
2. `*_saturate_to_zero` saturates at 0 for the minimum bound instead of i32::MIN

The function names could probably use some work, I had a difficult time naming them in a way that wasn't too verbose while also being distinct from the builtin Rust `.saturating_*` methods.

### CODE REVIEW

File: extension/src/saturation.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
+use pgx::*;
+
+/// Saturating Math for Integers
+
+/// Computes x+y, saturating at the numeric bounds instead of overflowing
+#[pg_extern(schema = "toolkit_experimental", immutable, parallel_safe)]
+fn addition_saturate_to_min(x: i32, y: i32) -> i32 {
```
Comment (WireBaron) [inline extension/src/saturation.rs:7]: I'd go with `saturating_add` here.  It's more succinct and makes it more obvious that it's saturating on both high and low ends.

File: extension/src/saturation.rs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    x.saturating_add(y)
+}
+
+/// Computes x+y, saturating at 0 for the minimum bound instead of i32::MIN
+#[pg_extern(schema = "toolkit_experimental", immutable, parallel_safe)]
+fn addition_saturate_to_zero(x: i32, y: i32) -> i32 {
```
Comment (WireBaron) [inline extension/src/saturation.rs:13]: Maybe `saturating_add_pos` or `pos_saturating_add`?

File: extension/src/saturation.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+/// Computes x*y, saturating at the numeric bounds instead of overflowing
+#[pg_extern(schema = "toolkit_experimental", immutable, parallel_safe)]
+fn multiplication_saturate_to_min(x: i32, y: i32) -> i32 {
+    x.saturating_mul(y)
+}
+
+/// Computes x*y, saturating at 0 for the minimum bound instead of i32::MIN
+#[pg_extern(schema = "toolkit_experimental", immutable, parallel_safe)]
+fn multiplication_saturate_to_zero(x: i32, y: i32) -> i32 {
```
Comment (WireBaron) [inline extension/src/saturation.rs:47]: I don't think we actually need a saturate_to_zero for multiplication.  If a user needs this to be positive they probably need the inputs to this to be positive.  If they've checked that, then the result will necessarily be positive.

File: extension/src/saturation.rs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    if result > 0 {
+        result
+    } else {
+        0
+    }
+}
+
+/// Computes x-y, saturating at the numeric bounds instead of overflowing.
+#[pg_extern(schema = "toolkit_experimental", immutable, parallel_safe)]
+fn subtraction_saturate_to_min(x: i32, y: i32) -> i32 {
```
Comment (WireBaron) [inline extension/src/saturation.rs:25]: `saturating_sub`, and `pos_saturating_sub`?

**Review by WireBaron — COMMENTED** [review]
Code looks good, but we do need to figure out better names.  Maybe @rtwalker has some good suggestions?

Also adding @JLockerman as a reviewer since he requested this.

**Review by epgts — COMMENTED** [review]
I like Brian's names. Otherwise LGTM.

File: extension/src/saturation.rs (hunk 2 above)
Comment (rtwalker) [reply extension/src/saturation.rs:13]: +1 for `saturating_add_pos`

**Comment by thatzopoulos** [comment]
Updated names to `saturating_***` and `saturating_***_pos`. Removed test for multiplication saturating to zero.

**Review by rtwalker — APPROVED** [review]

**Review by WireBaron — APPROVED** [review]

**Pull request merged**